### DISCUSSION TITLE
Target-typed followed by a method returning the same type.

### DISCUSSION THREAD

**Discussion by iam3yal** [dotnet/csharplang#6942] [Language Ideas]
I'm not sure I've seen an issue about it but say I have the following scenario:
```csharp
    protected override AttributeBuilder GetExtendedInputAttributes()
        => new AttributeBuilder()
            .Add("type", InputType.Text)
            .Add("autocomplete", AutoComplete)
            .Add("required", Required)
            .Add("dirname", DirectionName)
            .Add("readonly", IsReadOnly)
            .Add("placeholder", Placeholder)
            .Add("list", DataList)
            .Add("pattern", Pattern)
            .Add("minlength", MinLength)
            .Add("maxlength", MaxLength);
```
I know that the following works but not when it follows a method that returns the same type.
```csharp
protected override AttributeBuilder GetExtendedInputAttributes()
        => new ();
```
What I'd like to do is something like this:
```csharp
protected override AttributeBuilder GetExtendedInputAttributes()
        => new ().Add("type", InputType.Text);
```

**Comment by HaloFour** [comment]
It's unrelated to the return type of the method, target-typed returns can't negotiate a fluent chain in general:

```cs
AttributeBuilder builder = new ()
        .Add("type", InputType.Text)
        .Add("autocomplete", AutoComplete)
        .Add("required", Required)
        .Add("dirname", DirectionName)
        .Add("readonly", IsReadOnly)
        .Add("placeholder", Placeholder)
        .Add("list", DataList)
        .Add("pattern", Pattern)
        .Add("minlength", MinLength)
        .Add("maxlength", MaxLength);
```

I want to say there was a discussion on this subject before but my search-fu has failed me this morning.

**Reply by HaloFour** [reply]
I'm, personally, not a huge fan of inference at a distance like this.  Even if the rules were strict and the only methods that could be allowed to match would be instance and extension methods where the instance type and the return type were the same I think it makes the code a bit more difficult to reason about.  I think it would also be more fragile to changes in the code which could break the inference chain.

**Reply by iam3yal** [reply]
In my case implementing `IEnumerable` is not desirable on this specific type but I could potentially do this and then I could use collection initializer like this.
```csharp
protected override AttributeBuilder GetExtendedInputAttributes()
    => new ()
    {
        {"type", InputType.Text}, 
        // ...
    };
```
Not to mention that I have a special `Add` method where I could pass an expression to conditionally add the attribute which in this case a collection initializer wouldn't work at all.

Why inference should work for collection initializers that in fact use `Add` under the hood but not other methods that return the same type?

**Reply by TahirAhmadov** [reply]
Yes, this was discussed explicitly. The problem is that if this is attempted to be supported, then when the compiler encounters the last `.Add()` call, it has to work backwards, and there are hundreds of `Add` methods in any given project, including BCL and custom code. Not only it's prohibitively expensive, but it can also lead to ambiguity - what if multiple `Add` methods satisfy the criteria?

**Reply by HaloFour** [reply]
Right, I don't think it could be remotely as flexible as any fluent chain where the intermediate values could be of any type.  I think it would only work if every method in the chain returned the same type as the target type.  So given:

```cs
MyClass result = new()
    .Foo("123")
    .Bar(123);
```

The compiler would look for an instance/extension method of signature `MyClass Bar(string)` and then look for an instance/extension method of signature `MyClass Foo(int)`.

Which isn't to say that it's too niche to be useful, but it sounds like it'd be pretty easy to break the chain of inference.

**Reply by TahirAhmadov** [reply]
Even if compiler were to look for `MyClass Bar(string)`, it still leaves open the question - where to look for it? And what if there are more than one method which satisfy this signature?
The only workable solution would be a "special case" construct - when the compiler sees `new().Foo("123").Bar(123)`, it switches to a special mode, where it infers that `new()` means `MyClass`, and attempts to resolve the methods that way. This looks like too much of a "departure" from the normal C# compilation "path".

**Reply by HaloFour** [reply]
> where to look for it?

As an instance method of `MyClass` or an extension method that accepts `MyClass`.

>  it switches to a special mode, where it infers that `new()` means MyClass, and attempts to resolve the methods that way.

Which would have the same end result and would probably be easier to implement.

**Reply by iam3yal** [reply]
@TahirAhmadov Not sure where you're taking your facts from but this isn't new problem or abnormal.

**Reply by TahirAhmadov** [reply]
Exactly, the only way would be for the compiler to take a look at the whole expression and check if it matches a specific pattern, in order to know that `MyClass` is the type to work with. And I'm not aware of any other place where C# (or other languages, for that matter) do something like this.
PS. What I mean is, the compiler encounters `new()`, and then has to perform a very complex look-ahead operation, with potentially many steps, in order to figure out, whether this is legal code or not. So far, I heard that even limited look-aheads are not viewed with much excitement because of the complexity; here, the compiler would need to potentially parse hundreds of tokens, in order to determine whether the whole thing compiles or not.

**Reply by iam3yal** [reply]
@TahirAhmadov Why? it's just need to look to see whether the last call returns the same type, why would it have to go through all calls? it doesn't make sense to me.

**Reply by HaloFour** [reply]
@eyalalonn 

> Why? it's just need to look to see whether the last call returns the same type, why would it have to go through all calls? it doesn't make sense to me.

It can't even resolve the last call unless it knows the target type of each call within the chain.  It would be impossible for the compiler to handle this inference in reverse via normal means because those methods could return any intermediate types, so this feature would require that the compiler use a very different mechanism of inference which puts it on pretty tight rails.

**Reply by AartBluestoke** [reply]
could the compiler short-circuit at a stack of calls that look like

```
Class Foo{
Foo Bar();
Foo Baz();
Foo Zod()
}

Foo f;

var f.Baz().Baz().Zod().Bar().Zod();
```
the compiler should ( in the circumstances that a function call on Type T returns a type T) know the type resolution doesn't change?
(but that makes this a compiler question as much as a language question)

**Comment by iam3yal** [comment]
Well, one way to *solve* it is to do something like this:
```csharp
protected override AttributeBuilder GetExtendedInputAttributes()
    => new (builder => builder
            .Add(/* ... */)
            .Add(/* ... */));
```
But if we had something like an implicit parameter similar to Kotlin it could have been a bit nicer. 
```csharp
protected override AttributeBuilder GetExtendedInputAttributes()
    => new (it.Add(/* ... */)
              .Add(/* ... */));
```
Just as a side note in Kotlin it's more idiomatic to use type-safe builders for this so it would look something like this:
```csharp
fun foo() = AttributeBuilder.build {
    add(/* ... */)
    add(/* ... */)
}
```

**Comment by svick** [comment]
Previous discussion about this: https://github.com/dotnet/csharplang/discussions/5211.
